### PR TITLE
Intruduce changesets for associations

### PR DESCRIPTION
**Pull request by michalmuskala**

### CODE REVIEW

File: lib/ecto/changeset/relation.ex [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  def change(_related, nil, nil), do: {:ok, nil, false, true}
+
+  def change(%{related: mod} = related, value, current) do
+    {pk, _} = primary_key(mod)
+    cast_or_change(related, value, current, pk, pk, &do_change(&1, &2, mod))
+  end
+
+  defp do_change(value, nil, mod) do
+    fields    = mod.__schema__(:fields)
+    embeds    = mod.__schema__(:embeds)
```
Comment (josevalim) [inline lib/ecto/changeset/relation.ex]: We will need to include `:assocs` here?

File: lib/ecto/changeset.ex [hunk 2]
```diff
@@ -729,6 +738,9 @@ defmodule Ecto.Changeset do
         case Map.get(types, key) do
           {:embed, embed} ->
             {key, Ecto.Embedded.apply_changes(embed, value)}
+          {:assoc, _assoc} ->
+            raise ArgumentError, "associations should be removed from changeset" <>
```
Comment (josevalim) [inline lib/ecto/changeset.ex]: What is the idea here?

Comment (michalmuskala) [inline lib/ecto/changeset.ex]: We will be calling `Repo.insert!/update!/delete!` for the assocs, and using the returned model from there, and not applying the changes from the changeset (repo functions do that already for us)

**Comment by josevalim** [comment]
Really nice. I think we should also port some of the tests from test/ecto/embedded_test.exs to test/ecto/association_test.exs. You should probably change the latter first though to make sure they use the same model names, it will help when moving things over and finding corner cases.

File: lib/ecto/changeset.ex (hunk 2 above)
Comment (josevalim) [inline lib/ecto/changeset.ex]: I see, perfect.

**Comment by michalmuskala** [comment]
Yes, we will need the tests. Making this thing actually work is the nest step ;)
For now it's mainly renaming.

**Comment by paulcsmith** [comment]
Thanks for your work @michalmuskala This is going to be really helpful :D

File: lib/ecto/changeset.ex (hunk 2 above)
Comment (sumerman) [inline lib/ecto/changeset.ex]: We are using `apply_changes` to build a preview before saving into the DB. Right now, with no nested changesets support we have to manually propagate application in order to build this preview. It would be sad to keep this mess even when Ecto will be able to handle this for saving.

Comment (josevalim) [inline lib/ecto/changeset.ex]: @sumerman are you advocating for `apply_changes` to also apply the association changes?

Comment (sumerman) [inline lib/ecto/changeset.ex]: @josevalim yes, it makes perfect sense for me. /cc @michalmuskala

Comment (josevalim) [inline lib/ecto/changeset.ex]: If so, this would definitely be possible to do, we just need to make sure we strip all associations from changes before we call `apply_changes/1` inside the repository.

Comment (michalmuskala) [inline lib/ecto/changeset.ex]: Yeah, we could definitely do that, and the code is mostly there with embedded models.

Comment (sumerman) [inline lib/ecto/changeset.ex]: @michalmuskala that would be great! thanks in advance.

**Comment by michalmuskala** [comment]
@josevalim Ready for review!

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by gjaldon** [comment]
Great! Looking forward to using this :)
